minifi/minifi-nar-bundles/minifi-framework-bundle/minifi-framework/minifi-framework-core/src/main/java/org/apache/nifi/minifi/c2/command/syncresource/DefaultSyncResourceStrategy.java

```diff
@@ -0,0 +1,212 @@
+
+    private static final Set<Entry<OperationState, OperationState>> SUCCESS_RESULT_PAIRS = Set.of(
+        entry(NO_OPERATION, NO_OPERATION),
+        entry(NO_OPERATION, FULLY_APPLIED),
+        entry(FULLY_APPLIED, NO_OPERATION),
+        entry(FULLY_APPLIED, FULLY_APPLIED));
+
+    private static final Set<Entry<OperationState, OperationState>> FAILED_RESULT_PAIRS = Set.of(
+        entry(NO_OPERATION, NOT_APPLIED),
+        entry(NOT_APPLIED, NO_OPERATION),
+        entry(NOT_APPLIED, NOT_APPLIED));
+
+    private final ResourceRepository resourceRepository;
+    private final Path assetDirectory;
+    private final Path extensionDirectory;
+
+    public DefaultSyncResourceStrategy(ResourceRepository resourceRepository, Path assetDirectory, Path extensionDirectory) {
+        this.resourceRepository = resourceRepository;
+        this.assetDirectory = assetDirectory;
+        this.extensionDirectory = extensionDirectory;
+    }
+
+    @Override
+    public OperationState synchronizeResourceRepository(ResourcesGlobalHash c2GlobalHash, List<ResourceItem> c2ServerItems,
+                                                        BiFunction<String, Function<InputStream, Optional<Path>>, Optional<Path>> resourceDownloadFunction,
+                                                        Function<String, Optional<String>> urlEnrichFunction) {
+        Set<ResourceItem> c2Items = c2ServerItems.stream().collect(toSet());
+        Set<ResourceItem> agentItems = resourceRepository.findAllResourceItems().stream().collect(toSet());
+
+        OperationState deleteResult = deleteItems(c2Items, agentItems);
+        OperationState saveResult = saveNewItems(c2Items, agentItems, resourceDownloadFunction, urlEnrichFunction);
+
+        Entry<OperationState, OperationState> resultPair = entry(deleteResult, saveResult);
+
+        return SUCCESS_RESULT_PAIRS.contains(resultPair)
+            ? saveGlobalHash(c2GlobalHash, deleteResult, saveResult)
+            : FAILED_RESULT_PAIRS.contains(resultPair) ? NOT_APPLIED : PARTIALLY_APPLIED;
+    }
+
+    private OperationState deleteItems(Set<ResourceItem> c2Items, Set<ResourceItem> agentItems) {
+        List<ResourceItem> toDeleteItems = agentItems.stream().filter(not(c2Items::contains)).toList();
+        if (toDeleteItems.isEmpty()) {
+            return NO_OPERATION;
+        }
+
+        List<ResourceItem> deletedItems = toDeleteItems.stream()
+            .map(this::deleteItem)
+            .flatMap(Optional::stream)
+            .toList();
+
+        return deletedItems.isEmpty()
+            ? NOT_APPLIED
+            : deletedItems.size() == toDeleteItems.size() ? FULLY_APPLIED : PARTIALLY_APPLIED;
+    }
+
+    private Optional<ResourceItem> deleteItem(ResourceItem item) {
+        try {
+            Path resourcePath = resourcePath(item);
+            ResourceItem deletedItem = resourceRepository.deleteResourceItem(item);
+            deleteSilently(resourcePath, "Unable to delete resource file");
```

Comment:
Why don't we encapsulate all the delete logic in the repository's deleteResourceItem method - same for the add.
Maybe I misunderstood the intent of the FileResourceRepository. Is it only the FileMetadata Repository?

Reply:
Good point again. Initially I wanted it to be Metadata repository only. However with encapsulating the file handling logic it seems to be more coherent,